File: src-tauri/src/game_frames.js
/*
 * Preamble for the in-frame bridge the desktop app installs natively (see game_frames.rs).
 *
 * WebKitGTK runs this at document start in every frame of the app's webview, before any
 * of the frame's own script, including cross-origin game frames the app's JavaScript can
 * never reach. It is only installed while a game is being launched or played, and the Rust
 * side calls this function with that game's context, then runs the bridge itself only when
 * it returns a role.
 *
 * It must be cheap and must never throw into the page, and it decides very little on its
 * own: whether this frame is one of the app's game frames, and which part it plays.
 *
 *   game    the outermost frame the app did not serve itself: the game, or the portal
 *           shell that hosts it. Gets the full bridge (virtual storage, key detection,
 *           console input, pause, audio) under the game's catalog id.
 *   nested  a frame inside the game frame (the playable canvas of a portal shell, an SDK
 *           frame). Gets the bridge without a game id — focus and pause handling, but no
 *           storage of its own — plus console input relayed down from the game frame.
 *           Only one frame per game may own the saved profile: the bridge restores the
 *           freshest bucket it finds, so two owners would trade each other's saves.
 *   ad      a known ad or measurement host. Left alone.
 *
 * Everything else — the app's own top document, its same-origin shells and blob
 * documents, a puller relay page — is not touched here: those documents bring their own
 * bridge in their HTML, or are same-origin and reachable from the app directly.
 */
// eslint-disable-next-line @typescript-eslint/no-unused-vars -- called by name from game_frames.rs
function ptGameFramePreamble(CTX) {
	if (window.top === window) return null;

	var ancestors = location.ancestorOrigins;
	/* No ancestor list means no way to tell whose frame this is; stay out of it. */
	if (!ancestors || !ancestors.length) return null;
	if (ancestors[ancestors.length - 1] !== CTX.appOrigin) return null;
	if (location.protocol !== 'https:' && location.protocol !== 'http:') return null;
	if (location.origin === CTX.appOrigin) return null;
	for (var own = 0; own < CTX.ownOrigins.length; own++) {
		if (location.origin === CTX.ownOrigins[own]) return null;
	}

	/* document.open() re-runs user scripts against the same global. */
	if (window.__ptNativeGameFrame) return null;

	var AD_HOST =
		/(^|\.)(doubleclick\.net|googlesyndication\.com|googletagservices\.com|googletagmanager\.com|google-analytics\.com|googleadservices\.com|imasdk\.googleapis\.com|amazon-adsystem\.com|criteo\.com|pubmatic\.com|adnxs\.com|moatads\.com|rubiconproject\.com|openx\.net|casalemedia\.com|3lift\.com|taboola\.com|outbrain\.com|sentry\.io|hotjar\.com|facebook\.com|facebook\.net)$/;

	var outermost = true;
	for (var a = 0; a < ancestors.length - 1; a++) {
		if (ancestors[a] !== CTX.appOrigin) {
			outermost = false;
			break;
		}
	}

	var role = AD_HOST.test(location.hostname)
		? 'ad'
		: outermost && !CTX.topHasBridge
			? 'game'
			: 'nested';
	window.__ptNativeGameFrame = { role: role, gameId: role === 'game' ? CTX.gameId : '' };
	if (role === 'ad') return null;

	if (role === 'game') {
		/*
		 * First launch of this game with the bridge on this origin: start its virtual store
		 * from what the game already saved here while it ran without one, instead of from
		 * nothing. The saved profile, when the app has one, still takes precedence.
		 */
		try {
			var ns = '__pt_vs:' + CTX.gameId + ':';
			var real = window.localStorage;
			if (real.getItem(ns + 'ls') === null) {
				var seed = {};
				var seeded = 0;
				for (var i = 0; i < real.length; i++) {
					var key = real.key(i);
					if (!key || key.indexOf('__pt_vs:') === 0) continue;
					seed[key] = real.getItem(key);
					seeded++;
				}
				if (seeded) real.setItem(ns + 'ls', JSON.stringify(seed));
			}
		} catch {
			/* storage blocked for this origin — the bridge copes on its own */
		}
		window.__ptGameId = CTX.gameId;
		/*
		 * Proof of life for the launch watchdog: a frame that fires `load` without ever
		 * saying this never ran a script — a host that refused to be framed, an error page,
		 * a file that is not a page — and the app moves on to the next way to play it.
		 */
		try {
			window.top.postMessage(
				{
					type: 'potato-tomato-game-frame',
					role: role,
					gameId: CTX.gameId,
					href: String(location.href).slice(0, 300)
				},
				'*'
			);
		} catch {
			/* top unreachable */
		}
	} else {
		window.__ptGameId = '';
		/*
		 * Console input reaches the game frame by postMessage and its bridge forwards it to
		 * every child frame. Without a game id the bridge here does not listen, so dispatch it
		 * into this document and pass it further down.
		 */
		(function relayConsoleInput() {
			var KEY = {
				ArrowUp: 'ArrowUp',
				ArrowDown: 'ArrowDown',
				ArrowLeft: 'ArrowLeft',
				ArrowRight: 'ArrowRight',
				Space: ' ',
				Enter: 'Enter',
				Escape: 'Escape',
				ShiftLeft: 'Shift',
				ControlLeft: 'Control',
				Tab: 'Tab',
				Backspace: 'Backspace'
			};
			var KEY_CODE = {
				ArrowLeft: 37,
				ArrowUp: 38,
				ArrowRight: 39,
				ArrowDown: 40,
				Space: 32,
				Enter: 13,
				Escape: 27,
				ShiftLeft: 16,
				ControlLeft: 17,
				Tab: 9,
				Backspace: 8
			};
			var held = Object.create(null);
			var add = EventTarget.prototype.addEventListener;

			function keyOf(code) {
				if (KEY[code]) return KEY[code];
				if (/^Key[A-Z]$/.test(code)) return code.charAt(3).toLowerCase();
				if (/^Digit[0-9]$/.test(code)) return code.charAt(5);
				return code;
			}
			function keyCodeOf(code) {
				if (KEY_CODE[code] != null) return KEY_CODE[code];
				if (/^(Key[A-Z]|Digit[0-9])$/.test(code)) return code.charCodeAt(code.length - 1);
				return 0;
			}
			function target() {
				var active = document.activeElement;
				if (active && active !== document.body && active.tagName !== 'IFRAME') return active;
				return document.querySelector('canvas') || document.body || document.documentElement;
			}
			function dispatch(type, code) {
				var keyCode = keyCodeOf(code);
				var event;
				try {
					event = new KeyboardEvent(type, {
						key: keyOf(code),
						code: code,
						keyCode: keyCode,
						which: keyCode,
						bubbles: true,
						cancelable: true,
						composed: true,
						view: window
					});
					Object.defineProperty(event, 'keyCode', {
						get: function () {
							return keyCode;
						}
					});
					Object.defineProperty(event, 'which', {
						get: function () {
							return keyCode;
						}
					});
					target().dispatchEvent(event);
				} catch {
					/* a detached document mid-navigation */
				}
			}
			function forward(data) {
				var frames = document.getElementsByTagName('iframe');
				for (var f = 0; f < frames.length; f++) {
					try {
						if (frames[f].contentWindow) frames[f].contentWindow.postMessage(data, '*');
					} catch {
						/* frame gone */
					}
				}
			}
			add.call(window, 'message', function (event) {
				var data = event && event.data;
				if (!data || typeof data !== 'object' || data.type !== 'potato-tomato-touch-input') return;
				/* Only what comes down the frame chain from the app, never a sibling or a child. */
				if (event.source !== window.parent) return;
				var codes = Array.isArray(data.codes) ? data.codes : data.code ? [data.code] : [];
				var c;
				if (data.action === 'releaseAll') {
					for (c in held) dispatch('keyup', c);
					held = Object.create(null);
				} else if (data.action === 'down') {
					for (c = 0; c < codes.length; c++) {
						if (typeof codes[c] !== 'string' || held[codes[c]]) continue;
						held[codes[c]] = true;
						dispatch('keydown', codes[c]);
					}
				} else if (data.action === 'up') {
					for (c = 0; c < codes.length; c++) {
						if (typeof codes[c] !== 'string' || !held[codes[c]]) continue;
						delete held[codes[c]];
						dispatch('keyup', codes[c]);
					}
				} else {
					return;
				}
				forward(data);
			});
		})();
	}
	return role;
}

File: src-tauri/src/game_frames.rs
//! Scripts the app runs inside game frames it cannot reach from JavaScript.
//!
//! A game embedded from its own host is a cross-origin document: the page that frames it
//! cannot read it, write to it or add a script to it. The webview can. On Linux the app
//! installs `static/game-storage-bridge.child.js` into the game's frames natively, as a
//! WebKitGTK user script, so virtual storage, key detection and the touch console work in
//! a game played straight from its own host — which is what the Node relay used to exist
//! for, at the price of re-fetching and rewriting every game through a proxy.
//!
//! The script is not permanent. The frontend calls `set_game_frame_context` while it
//! resolves a launch, before the frame exists, and the user script is swapped for one that
//! carries that game's id. A page with no game on it has no bridge in its frames at all.
//!
//! Android has its own document-start bridge (`native_touch_bridge.js`, installed by
//! `MainActivity`), so nothing here runs there.

// Only the Linux webview can install the script; elsewhere the commands answer "unsupported".
#![cfg_attr(not(target_os = "linux"), allow(dead_code))]

use serde::Deserialize;

/// The shared in-frame bridge, byte for byte what the web build serves.
pub(crate) const BRIDGE_SOURCE: &str = include_str!("../../static/game-storage-bridge.child.js");
/// Decides whether a frame is a game frame and which part it plays; see the file.
const PREAMBLE_SOURCE: &str = include_str!("game_frames.js");

/// What the frontend knows about the launch it is resolving.
#[derive(Debug, Clone, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct GameFrameContext {
  /// Catalog id; the bridge keys the game's saved profile on it.
  pub game_id: String,
  /// Origin of the app document hosting the frames (`tauri://localhost`, or the dev server).
  pub app_origin: String,
  /// Origins that serve documents with the bridge already in their HTML (a running puller).
  #[serde(default)]
  pub own_origins: Vec<String>,
  /// The frame the app creates holds an app-made document that carries the bridge (a blob
  /// shell, an offline mirror), so frames inside it must not claim the game's profile.
  #[serde(default)]
  pub top_has_bridge: bool,
}

/// Catalog ids as they occur (`crazygames-2048`, `minecraft-1.8.8`, `2016 - diamond run`),
/// and nothing that could name another directory: shared by every command and scheme that
/// turns an id into a path or a URL.
pub fn is_catalog_id(id: &str) -> bool {
  let mut chars = id.chars();
  matches!(chars.next(), Some(c) if c.is_ascii_alphanumeric())
    && id.len() <= 200
    && !id.contains("..")
    && id
      .chars()
      .all(|c| c.is_ascii_alphanumeric() || matches!(c, '-' | '_' | '.' | ' '))
}

/// `scheme://host[:port]` and nothing more — the value is compared to `location.origin`.
fn is_origin(value: &str) -> bool {
  let Some((scheme, rest)) = value.split_once("://") else {
    return false;
  };
  !scheme.is_empty()
    && scheme
      .chars()
      .all(|c| c.is_ascii_alphanumeric() || c == '+' || c == '-' || c == '.')
    && !rest.is_empty()
    && !rest.contains(['/', '?', '#', '@', ' '])
}

impl GameFrameContext {
  pub fn validate(&self) -> Result<(), String> {
    if !is_catalog_id(&self.game_id) {
      return Err(format!("not a catalog game id: {:?}", self.game_id));
    }
    for origin in std::iter::once(&self.app_origin).chain(self.own_origins.iter()) {
      if !is_origin(origin) {
        return Err(format!("not an origin: {origin:?}"));
      }
    }
    Ok(())
  }
}

/// The user script for one launch: the context, the preamble, then the bridge — which
/// runs only in frames the preamble gives a role.
pub fn frame_script(context: &GameFrameContext) -> String {
  let ctx = serde_json::json!({
    "gameId": context.game_id,
    "appOrigin": context.app_origin,
    "ownOrigins": context.own_origins,
    "topHasBridge": context.top_has_bridge,
  });
  let mut out = String::with_capacity(BRIDGE_SOURCE.len() + PREAMBLE_SOURCE.len() + 512);
  out.push_str("(function () {\nvar ptRole;\ntry {\nptRole = (");
  out.push_str(PREAMBLE_SOURCE);
  out.push_str(")(");
  out.push_str(&ctx.to_string());
  out.push_str(");\n} catch (e) {\nreturn;\n}\nif (!ptRole) return;\n");
  out.push_str(BRIDGE_SOURCE);
  out.push_str("\n})();\n");
  out
}

#[cfg(target_os = "linux")]
mod linux {
  use std::cell::RefCell;
  use std::collections::HashMap;
  use webkit2gtk::{
    UserContentInjectedFrames, UserContentManagerExt, UserScript, UserScriptInjectionTime,
    WebViewExt,
  };

  thread_local! {
    /// The active script per webview label. Touched only on the GTK main thread, which is
    /// where `with_webview` runs its closure.
    static ACTIVE: RefCell<HashMap<String, UserScript>> = RefCell::new(HashMap::new());
  }

  /// Replace this webview's game-frame script; `None` just removes it.
  pub fn swap(webview: &tauri::Webview, source: Option<String>) -> Result<(), String> {
    let label = webview.label().to_string();
    let (done_tx, done_rx) = std::sync::mpsc::channel();
    webview
      .with_webview(move |platform| {
        let view = platform.inner();
        let Some(manager) = view.user_content_manager() else {
          let _ = done_tx.send(Err("webview has no user content manager".to_string()));
          return;
        };
        ACTIVE.with(|active| {
          let mut active = active.borrow_mut();
          if let Some(previous) = active.remove(&label) {
            manager.remove_script(&previous);
          }
          if let Some(source) = source {
            let script = UserScript::new(
              &source,
              UserContentInjectedFrames::AllFrames,
              UserScriptInjectionTime::Start,
              &[],
              &[],
            );
            manager.add_script(&script);
            active.insert(label, script);
          }
        });
        let _ = done_tx.send(Ok(()));
      })
      .map_err(|e| e.to_string())?;
    /*
     * Wait for the main thread to have done it: the frontend sets the frame's src as soon
     * as this command returns, and the frame's first document must already see the script.
     */
    done_rx
      .recv_timeout(std::time::Duration::from_secs(3))
      .map_err(|_| "timed out installing the game-frame script".to_string())?
  }
}

/// Whether this build can put the bridge into cross-origin game frames itself.
#[tauri::command]
pub fn native_game_frames_supported() -> bool {
  cfg!(target_os = "linux")
}

/// Install the in-frame bridge for the game about to launch in this webview.
///
/// Returns `false` where the platform has no way to do it (the frontend then keeps its
/// same-origin fallbacks), `true` once the script is in place for the next frame load.
#[tauri::command]
pub async fn set_game_frame_context(
  webview: tauri::Webview,
  context: GameFrameContext,
) -> Result<bool, String> {
  context.validate()?;
  #[cfg(target_os = "linux")]
  {
    let source = frame_script(&context);
    tauri::async_runtime::spawn_blocking(move || linux::swap(&webview, Some(source)))
      .await
      .map_err(|e| e.to_string())??;
    Ok(true)
  }
  #[cfg(not(target_os = "linux"))]
  {
    let _ = webview;
    Ok(false)
  }
}

/// Take the bridge back out of this webview's frames (no game on screen).
#[tauri::command]
pub async fn clear_game_frame_context(webview: tauri::Webview) -> Result<(), String> {
  #[cfg(target_os = "linux")]
  {
    tauri::async_runtime::spawn_blocking(move || linux::swap(&webview, None))
      .await
      .map_err(|e| e.to_string())??;
  }
  #[cfg(not(target_os = "linux"))]
  let _ = webview;
  Ok(())
}

/// Debug builds only: `POTATO_TOMATO_FRAME_PROBE=<path to a .js file>` runs that file at
/// document start in every frame, for launch measurements (`scripts/play-path-bench.mjs`
/// passes `scripts/play-path/frame-probe.js`). A release build never reads the variable.
pub fn frame_probe_plugin<R: tauri::Runtime>() -> Option<tauri::plugin::TauriPlugin<R>> {
  if !cfg!(debug_assertions) {
    return None;
  }
  let path = std::env::var_os("POTATO_TOMATO_FRAME_PROBE")?;
  match std::fs::read_to_string(&path) {
    Ok(source) => Some(
      tauri::plugin::Builder::new("frame-probe")
        .js_init_script_on_all_frames(source)
        .build(),
    ),
    Err(e) => {
      eprintln!("POTATO_TOMATO_FRAME_PROBE: could not read {path:?}: {e}");
      None
    }
  }
}

#[cfg(test)]
mod tests {
  use super::*;

  fn ctx() -> GameFrameContext {
    GameFrameContext {
      game_id: "crazygames-2048".into(),
      app_origin: "tauri://localhost".into(),
      own_origins: vec!["http://127.0.0.1:18787".into()],
      top_has_bridge: false,
    }
  }

  #[test]
  fn accepts_catalog_ids_and_origins() {
    assert!(ctx().validate().is_ok());
    let mut dev = ctx();
    dev.app_origin = "http://127.0.0.1:5177".into();
    dev.game_id = "002".into();
    assert!(dev.validate().is_ok());
  }

  #[test]
  fn rejects_anything_that_is_not_an_id_or_an_origin() {
    for bad in [
      "",
      "../x",
      "x\"y",
      "a/b",
      "</script>",
      ".hidden",
      "a..b",
      "a\\b",
    ] {
      let mut c = ctx();
      c.game_id = bad.into();
      assert!(c.validate().is_err(), "{bad:?} accepted as a game id");
    }
    for bad in [
      "tauri://localhost/",
      "http://x/y",
      "localhost",
      "http://a@b",
      "javascript:1",
    ] {
      let mut c = ctx();
      c.app_origin = bad.into();
      assert!(c.validate().is_err(), "{bad:?} accepted as an origin");
    }
  }

  #[test]
  fn script_carries_context_as_json_before_the_bridge() {
    let script = frame_script(&ctx());
    let ctx_at = script
      .find("\"gameId\":\"crazygames-2048\"")
      .expect("context present");
    let bridge_at = script
      .find("potato-tomato-game-storage")
      .expect("bridge present");
    assert!(ctx_at < bridge_at);
    assert!(script.contains("function ptGameFramePreamble(CTX)"));
    assert!(script.contains("\"topHasBridge\":false"));
    assert!(script.trim_end().ends_with("})();"));
  }
}
